### PR TITLE
New videodock

### PR DESCRIPTION
**Pull request by jtpastro**
Feature #1211 - Prioritization of a video window of video dock on mouse click.

### CODE REVIEW

File: bigbluebutton-client/src/org/bigbluebutton/modules/videoconf/views/GraphicsWrapper.as [hunk 1]
```diff
@@ -37,10 +41,10 @@ package org.bigbluebutton.modules.videoconf.views
             return result;
         }
 
-        private function calculateCellDimensions(numColumns:int, numRows:int, cellAspectRatio:Number):Object {
+        private function calculateCellDimensions(canvasWidth:int, canvasHeight:int, numColumns:int, numRows:int, cellAspectRatio:Number):Object {
             var obj:Object = {
-                width: Math.floor(width / numColumns),
-                height: Math.floor(height / numRows)
+width: Math.floor(canvasWidth / numColumns),
```
Comment (fcecagno) [inline bigbluebutton-client/src/org/bigbluebutton/modules/videoconf/views/GraphicsWrapper.as]: Fix indentation.

File: bigbluebutton-client/src/org/bigbluebutton/modules/videoconf/views/GraphicsWrapper.as [hunk 2]
```diff
@@ -59,17 +63,17 @@ package org.bigbluebutton.modules.videoconf.views
             return obj;
         }
 
-        private function findBestConfiguration():Object {
+        private function findBestConfiguration(canvasWidth:int, canvasHeight:int, numChildrenInCanvas:int):Object {
             var cellAspectRatio:Number = minContentAspectRatio;
 
             var bestConfiguration:Object = {
-                occupiedArea: 0
+occupiedArea: 0
```
Comment (fcecagno) [inline bigbluebutton-client/src/org/bigbluebutton/modules/videoconf/views/GraphicsWrapper.as]: Fix indentation.

File: bigbluebutton-client/src/org/bigbluebutton/modules/videoconf/views/GraphicsWrapper.as [hunk 3]
```diff
@@ -122,8 +196,9 @@ package org.bigbluebutton.modules.videoconf.views
             var graphic:UserGraphicHolder = new UserGraphicHolder();
             graphic.userId = userId;
             graphic.addEventListener(FlexEvent.CREATION_COMPLETE, function(event:FlexEvent):void {
-                graphic.loadAvatar(_options);
-            });
+                    graphic.loadAvatar(_options);
```
Comment (fcecagno) [inline bigbluebutton-client/src/org/bigbluebutton/modules/videoconf/views/GraphicsWrapper.as]: Fix indentation.

File: bigbluebutton-client/src/org/bigbluebutton/modules/videoconf/views/GraphicsWrapper.as [hunk 4]
```diff
@@ -132,8 +207,9 @@ package org.bigbluebutton.modules.videoconf.views
             var graphic:UserGraphicHolder = new UserGraphicHolder();
             graphic.userId = userId;
             graphic.addEventListener(FlexEvent.CREATION_COMPLETE, function(event:FlexEvent):void {
-                graphic.loadVideo(_options, connection, streamName);
-            });
+                    graphic.loadVideo(_options, connection, streamName);
```
Comment (fcecagno) [inline bigbluebutton-client/src/org/bigbluebutton/modules/videoconf/views/GraphicsWrapper.as]: Fix indentation.

File: bigbluebutton-client/src/org/bigbluebutton/modules/videoconf/views/GraphicsWrapper.as [hunk 5]
```diff
@@ -162,11 +238,25 @@ package org.bigbluebutton.modules.videoconf.views
             var graphic:UserGraphicHolder = new UserGraphicHolder();
             graphic.userId = userId;
             graphic.addEventListener(FlexEvent.CREATION_COMPLETE, function(event:FlexEvent):void {
-                graphic.loadCamera(_options, camIndex, videoProfile);
-            });
+                    graphic.loadCamera(_options, camIndex, videoProfile);
+                    });
```
Comment (fcecagno) [inline bigbluebutton-client/src/org/bigbluebutton/modules/videoconf/views/GraphicsWrapper.as]: Fix indentation.

File: bigbluebutton-client/src/org/bigbluebutton/modules/videoconf/model/VideoConfOptions.as [hunk 6]
```diff
@@ -162,6 +164,9 @@ package org.bigbluebutton.modules.videoconf.model
 				if (vxml.@glowBlurSize != undefined) {
 					glowBlurSize = Number(vxml.@glowBlurSize.toString());
 				}
+				if (vxml.@priorityRatio != undefined) {
+                    priorityRatio = Number(vxml.@priorityRatio.toString());
+				}
 			}
```
Comment (fcecagno) [inline bigbluebutton-client/src/org/bigbluebutton/modules/videoconf/model/VideoConfOptions.as:170]: First, please be careful with the indentation. It should always follow the pattern used in the file (spaces or tabs) - you don't need to fix any indentation from others, but the new code should always be indented properly.

Second, when you add a new option such as priorityRatio, you should include that option in the config.xml.template (https://github.com/jtpastro/bigbluebutton/blob/new-videodock/bigbluebutton-client/resources/config.xml.template#L56), so others don't need to look into the code to be aware of a given module options - they just need to check the config template.

File: bigbluebutton-client/src/org/bigbluebutton/modules/videoconf/views/GraphicsWrapper.as [hunk 7]
```diff
@@ -134,7 +225,10 @@ package org.bigbluebutton.modules.videoconf.views
             graphic.addEventListener(FlexEvent.CREATION_COMPLETE, function(event:FlexEvent):void {
                 graphic.loadVideo(_options, connection, streamName);
             });
-            super.addChild(graphic);
+            graphic.addEventListener(MouseEvent.CLICK, onVBoxClick); 
+            super.addChild(graphic); 
+            if(graphic.contentAspectRatio < cellAspectRatio)
+                cellAspectRatio = graphic.contentAspectRatio;
```
Comment (fcecagno) [inline bigbluebutton-client/src/org/bigbluebutton/modules/videoconf/views/GraphicsWrapper.as:234]: Are you sure that, at this point, graphic has finished to load the video and has the contentAspectRatio properly set? Please notice that graphic.loadVideo is only called when the component is completely created (FlexEvent.CREATION_COMPLETE). I'm not 100% sure, but I think that this comparison and assignment should be made after graphic.loadVideo, inside the event listener.

The same applies to addCameraForHelper and addAvatarFor.

File: bigbluebutton-client/src/org/bigbluebutton/modules/videoconf/views/GraphicsWrapper.as [hunk 8]
```diff
@@ -164,7 +258,23 @@ package org.bigbluebutton.modules.videoconf.views
...
             });
+
+            graphic.addEventListener(MouseEvent.CLICK, onVBoxClick);
             super.addChild(graphic);
+            if(graphic.contentAspectRatio < cellAspectRatio)
+                cellAspectRatio = graphic.contentAspectRatio;
+        }
+
+        protected function onVBoxClick(event:MouseEvent = null):void {
+            var item:UserGraphicHolder = event.currentTarget as UserGraphicHolder;
```
Comment (fcecagno) [inline bigbluebutton-client/src/org/bigbluebutton/modules/videoconf/views/GraphicsWrapper.as:274]: If you need event.currentTarget, event should never be null, so you should probably remove the " = null" from the function definition.

File: bigbluebutton-client/resources/config.xml.template [hunk 9]
```diff
@@ -1,6 +1,6 @@
-<?xml version="1.0" ?>
+<?XMl version="1.0" ?>
 <config>
-    <localeversion suppressWarning="false">0.8</localeversion>
+    <Localeversion suppressWarning="false">0.8</localeversion>
```
Comment (fcecagno) [inline bigbluebutton-client/resources/config.xml.template:3]: Is there any particular reason to change the case of these XML nodes?

Comment (jtpastro) [inline bigbluebutton-client/resources/config.xml.template:3]: No. I've probably done that by mistake.

**Pull request merged**